src/internal.c

```diff
@@ -15237,11 +15247,19 @@ static int DoCertificate(WOLFSSL* ssl, byte* input, word32* inOutIdx,
     WOLFSSL_ENTER("DoCertificate");
 
 #ifdef SESSION_CERTS
-    /* Reset the session cert chain count in case the session resume failed. */
-    ssl->session->chain.count = 0;
-    #ifdef WOLFSSL_ALT_CERT_CHAINS
+    /* Reset the session cert chain count in case the session resume failed,
+    do not reset if we are resuming after an async wait */
+#if defined(WOLFSSL_ASYNC_CRYPT) || defined(WOLFSSL_NONBLOCK_OCSP)
+    if (ssl->async == NULL ||
+        (((ProcPeerCertArgs*)(ssl->async->args))->lastErr != OCSP_WANT_READ &&
```

Comment:
Is it possible to use `ssl->error` for this? I believe it might still have the WC_PENDING_E or OSCP_WANT_READ.  I am not a fan of this async cast to ProcPeerCertArgs since it is not guaranteed to be that struct. Another option might be to not reset if ssl->async is populated or something like that?

Reply:
good point, I tested and ssl->error does hold the async errors so it works the same. Should I keep changes I made in setting lastErr and excluding async errors in TLS_ASYNC_FINALIZE just so lastErr doesn't have different values based on where the error occured?

Reply:
Yes revert the other changes if you can.